// @code-inspector/turbopack@1.5.1/dist/index.mjs
var a = Object.defineProperty, p = Object.defineProperties;
var d = Object.getOwnPropertyDescriptors;
var n = Object.getOwnPropertySymbols;
var f = Object.prototype.hasOwnProperty, j = Object.prototype.propertyIsEnumerable;
var i = (e, t, o) => t in e ? a(e, t, { enumerable: !0, configurable: !0, writable: !0, value: o }) : e[t] = o, r = (e, t) => {
  for (var o in t || (t = {}))
    f.call(t, o) && i(e, o, t[o]);
  if (n)
    for (var o of n(t))
      j.call(t, o) && i(e, o, t[o]);
  return e;
}, s = (e, t) => p(e, d(t));
import { isDev as v, isNextGET16 as m } from "@code-inspector/core";
import b from "path";
import { fileURLToPath as y } from "url";
function g(e) {
  if (typeof e.importMetaResolve == "function") {
    const t = e.importMetaResolve(
      "@code-inspector/webpack"
    );
    return y(t);
  }
  return typeof e.requireResolve == "function" ? e.requireResolve("@code-inspector/webpack") : null;
}
function D(e) {
  if (e.close || !v(e.dev, process.env.NODE_ENV === "development"))
    return {};
  const t = {
    port: 0,
    entry: "",
    output: e.output
  }, o = g({
    requireResolve: typeof require == "undefined" ? null : require.resolve,
    importMetaResolve: import.meta.resolve
  }), l = b.resolve(o, ".."), u = [
    "*.jsx",
    "*.tsx",
    "layout.js",
    "layout.ts",
    "page.js",
    "page.ts",
    "loading.js",
    "loading.ts",
    "not-found.js",
    "not-found.ts",
    "error.js",
    "error.ts",
    "global-error.js",
    "global-error.ts",
    "template.js",
    "template.ts",
    "default.js",
    "default.ts"
  ], c = "**/{".concat(u.join(","), "}");
  return {
    [m() ? "**/*.{jsx,tsx,js,ts,mjs,mts}" : c]: {
      loaders: [
        {
          loader: "".concat(l, "/loader.js"),
          options: s(r({}, e), {
            record: t
          })
        },
        {
          loader: "".concat(l, "/inject-loader.js"),
          options: s(r({}, e), {
            record: t
          })
        }
      ]
    }
  };
}
export {
  D as TurbopackCodeInspectorPlugin,
  g as resolveWebpackEntry
};
