// @code-inspector/webpack@1.5.1/dist/index.mjs
var N = Object.defineProperty, O = Object.defineProperties;
var R = Object.getOwnPropertyDescriptors;
var x = Object.getOwnPropertySymbols;
var S = Object.prototype.hasOwnProperty, z = Object.prototype.propertyIsEnumerable;
var W = (e, t, s) => t in e ? N(e, t, { enumerable: !0, configurable: !0, writable: !0, value: s }) : e[t] = s, c = (e, t) => {
  for (var s in t || (t = {}))
    S.call(t, s) && W(e, s, t[s]);
  if (x)
    for (var s of x(t))
      z.call(t, s) && W(e, s, t[s]);
  return e;
}, y = (e, t) => O(e, R(t));
var h = (e, t, s) => new Promise((o, n) => {
  var a = (r) => {
    try {
      f(s.next(r));
    } catch (i) {
      n(i);
    }
  }, u = (r) => {
    try {
      f(s.throw(r));
    } catch (i) {
      n(i);
    }
  }, f = (r) => r.done ? o(r.value) : Promise.resolve(r.value).then(a, u);
  f((s = s.apply(e, t)).next());
});
import { normalizePath as $, isDev as _, getProjectRecord as H, isNextjsProject as Q, getCodeWithWebComponent as U } from "@code-inspector/core";
import j from "path";
import { fileURLToPath as V } from "url";
function q(e, t) {
  return h(this, null, function* () {
    if (!e || !t)
      return [];
    const s = typeof e == "function" ? yield e() : e;
    let o = [];
    if (typeof s == "object" && !Array.isArray(s))
      for (const n in s) {
        const a = s[n], u = a.import || a;
        w(o, u, t);
      }
    else
      w(o, s, t);
    return o.filter((n) => !!n);
  });
}
function w(e, t, s) {
  typeof t == "string" ? e.push(D(t, s)) : Array.isArray(t) && e.push(
    ...t.map((o) => D(o, s))
  );
}
function D(e, t) {
  return j.isAbsolute(e) ? $(e) : e.startsWith(".") ? j.resolve(t, $(e)) : "";
}
const L = j.dirname(V(import.meta.url)), P = j.resolve(L, "./loader.js"), E = j.resolve(L, "./inject-loader.js");
function B(e) {
  return e.some(
    (t) => {
      var s, o;
      return (o = (s = t == null ? void 0 : t.use) == null ? void 0 : s.some) == null ? void 0 : o.call(
        s,
        (n) => (n == null ? void 0 : n.loader) === P || (n == null ? void 0 : n.loader) === E
      );
    }
  );
}
const F = (e, t) => {
  var a, u;
  const s = (t == null ? void 0 : t.compiler) || t, o = (a = s == null ? void 0 : s.options) == null ? void 0 : a.module, n = (o == null ? void 0 : o.rules) || (o == null ? void 0 : o.loaders) || [];
  B(n) || n.push(
    c({
      test: (u = e.match) != null ? u : /\.html$/,
      resourceQuery: /vue/,
      use: [
        {
          loader: P,
          options: e
        }
      ]
    }, e.enforcePre === !1 ? {} : { enforce: "pre" }),
    c({
      test: /\.(vue|jsx|tsx|js|ts|mjs|mts|svelte)$/,
      use: [
        {
          loader: P,
          options: e
        }
      ]
    }, e.enforcePre === !1 ? {} : { enforce: "pre" }),
    y(c({}, e.injectTo ? { resource: e.injectTo } : {
      test: /\.(jsx|tsx|js|ts|mjs|mts)$/,
      exclude: /node_modules/
    }), {
      use: [
        {
          loader: E,
          options: e
        }
      ],
      enforce: Q() ? "pre" : "post"
    })
  );
};
function T(e, t, s) {
  return U({
    options: y(c({}, e), { importClient: "code" }),
    file: "main.js",
    code: "",
    record: t,
    inject: !0,
    server: s
  });
}
function G(o) {
  return h(this, arguments, function* ({
    options: e,
    record: t,
    assets: s
  }) {
    const n = Object.keys(s).filter((a) => /\.html$/.test(a));
    if (n.length) {
      const a = yield T(e, t);
      n.forEach((u) => {
        var r, i;
        const f = (i = (r = s[u]) == null ? void 0 : r.source) == null ? void 0 : i.call(r);
        if (typeof f == "string") {
          const d = f.replace(
            "<head>",
            '<head><script type="module">\n'.concat(a, "\n<\/script>")
          );
          s[u] = {
            source: () => d,
            size: () => d.length
          };
        }
      });
    }
  });
}
class Y {
  constructor(t) {
    this.options = t;
  }
  apply(t) {
    return h(this, null, function* () {
      var n, a, u, f, r, i, d, b, k, A, p, C;
      if (this.options.close || !_(
        this.options.dev,
        ((n = t == null ? void 0 : t.options) == null ? void 0 : n.mode) === "development" || process.env.NODE_ENV === "development"
      ))
        return;
      const s = {
        port: 0,
        entry: "",
        output: this.options.output,
        inputs: q(
          (a = t == null ? void 0 : t.options) == null ? void 0 : a.entry,
          (u = t == null ? void 0 : t.options) == null ? void 0 : u.context
        ),
        envDir: (f = t == null ? void 0 : t.options) == null ? void 0 : f.context,
        root: (r = t == null ? void 0 : t.options) == null ? void 0 : r.context
      }, o = ((i = t == null ? void 0 : t.options) == null ? void 0 : i.cache) || ((b = (d = t == null ? void 0 : t.options) == null ? void 0 : d.experiments) == null ? void 0 : b.cache);
      if ((o == null ? void 0 : o.type) === "filesystem" && (this.options.cache ? (s.port = this.options.port || ((k = H(s)) == null ? void 0 : k.previousPort), T(this.options, s, !0)) : o.version = "code-inspector-".concat(Date.now())), F(y(c({}, this.options), { record: s }), t), (A = t == null ? void 0 : t.hooks) != null && A.emit && !((C = (p = this.options.skipSnippets) == null ? void 0 : p.includes) != null && C.call(p, "htmlScript"))) {
        const l = this.options;
        t.hooks.emit.tapAsync(
          "WebpackCodeInspectorPlugin",
          (v, I) => h(this, null, function* () {
            let g = {};
            v.getAssets ? g = yield v.getAssets() : g = v.assets, yield G({
              options: l,
              record: s,
              assets: g
            }), I();
          })
        );
      }
    });
  }
}
export {
  Y as default
};
