// @code-inspector/esbuild@1.5.1/dist/index.mjs
var f = (t, c, r) => new Promise((p, o) => {
  var u = (a) => {
    try {
      i(r.next(a));
    } catch (s) {
      o(s);
    }
  }, e = (a) => {
    try {
      i(r.throw(a));
    } catch (s) {
      o(s);
    }
  }, i = (a) => a.done ? p(a.value) : Promise.resolve(a.value).then(u, e);
  i((r = r.apply(t, c)).next());
});
import { isDev as g, getMappingFilePath as y, isExcludedFile as C, getCodeWithWebComponent as x, isJsTypeFile as T, transformCode as m, parseSFC as v } from "@code-inspector/core";
import w from "fs";
import F from "path";
const j = "@code-inspector/esbuild";
function E(t) {
  return {
    name: j,
    setup(c) {
      if (t.close || !g(t.dev, !1))
        return;
      const r = {
        port: 0,
        entry: "",
        output: t.output
      }, { escapeTags: p = [] } = t, o = /* @__PURE__ */ new Map();
      c.onLoad(
        { filter: t.match || /\.(jsx|tsx|js|ts|mjs|mts)?$/ },
        (u) => f(this, null, function* () {
          let e = u.path;
          e = y(e, t.mappings);
          const i = F.extname(e).replace(".", "");
          let a = yield w.promises.readFile(e, "utf8"), s = o.get(e);
          if (!s || s.originCode !== a) {
            let n = a;
            if (C(e, t))
              return {
                contents: n,
                loader: i
              };
            n = yield x({
              options: t,
              file: e,
              code: n,
              record: r
            });
            let l = "";
            if (T(e) ? l = "jsx" : e.endsWith(".svelte") && (l = "svelte"), l)
              n = yield m({
                content: n,
                filePath: e,
                fileType: l,
                escapeTags: p,
                pathType: t.pathType
              });
            else if (e.endsWith(".vue")) {
              l = "vue";
              const { descriptor: d } = v(n, {
                sourceMap: !1
              }), h = yield m({
                content: d.template.content,
                filePath: e,
                fileType: l,
                escapeTags: p,
                pathType: t.pathType
              });
              n = n.replace(d.template.content, h);
            }
            s = { originCode: a, output: { contents: n, loader: i } }, o.set(e, s);
          }
          return s.output;
        })
      );
    }
  };
}
export {
  E as EsbuildCodeInspectorPlugin
};
