// launch-ide@1.4.3/dist/index.mjs
var V = Object.defineProperty, W = Object.defineProperties;
var F = Object.getOwnPropertyDescriptors;
var N = Object.getOwnPropertySymbols;
var H = Object.prototype.hasOwnProperty, X = Object.prototype.propertyIsEnumerable;
var B = (t, e, n) => e in t ? V(t, e, { enumerable: !0, configurable: !0, writable: !0, value: n }) : t[e] = n, E = (t, e) => {
  for (var n in e || (e = {}))
    H.call(e, n) && B(t, n, e[n]);
  if (N)
    for (var n of N(e))
      X.call(e, n) && B(t, n, e[n]);
  return t;
}, P = (t, e) => W(t, F(e));
import T from "fs";
import v from "path";
import w from "child_process";
import Q from "os";
import l from "chalk";
import j from "dotenv";
const J = {
  "/Kiro.app/Contents/MacOS/Electron": "/Kiro.app/Contents/MacOS/Electron",
  "/Antigravity.app/Contents/MacOS/Electron": "/Antigravity.app/Contents/MacOS/Electron",
  "/Cursor.app/Contents/MacOS/Cursor": "/Cursor.app/Contents/MacOS/Cursor",
  "/Comate.app/Contents/MacOS/Electron": "/Comate.app/Contents/MacOS/Electron",
  "/Qoder.app/Contents/MacOS/Electron": "/Qoder.app/Contents/MacOS/Electron",
  "/Qoder CN.app/Contents/MacOS/Electron": "/Qoder CN.app/Contents/MacOS/Electron",
  "/Windsurf.app/Contents/MacOS/Electron": "/Windsurf.app/Contents/MacOS/Electron",
  "/Trae.app/Contents/MacOS/Electron": "/Trae.app/Contents/MacOS/Electron",
  "/Trae CN.app/Contents/MacOS/Electron": "/Trae CN.app/Contents/MacOS/Electron",
  "/Applications/CodeBuddy.app/Contents/MacOS/Electron": "/Applications/CodeBuddy.app/Contents/MacOS/Electron",
  "/Applications/CodeBuddy CN.app/Contents/MacOS/Electron": "/Applications/CodeBuddy CN.app/Contents/MacOS/Electron",
  "/Visual Studio Code.app/Contents/MacOS/Electron": "/Visual Studio Code.app/Contents/MacOS/Electron",
  "/Visual Studio Code.app/Contents/MacOS/Code": "/Visual Studio Code.app/Contents/MacOS/Code",
  "/Visual Studio Code - Insiders.app/Contents/MacOS/Electron": "/Visual Studio Code - Insiders.app/Contents/MacOS/Electron",
  "/VSCodium.app/Contents/MacOS/Electron": "/VSCodium.app/Contents/MacOS/Electron",
  "/WebStorm.app/Contents/MacOS/webstorm": "/WebStorm.app/Contents/MacOS/webstorm",
  "/HBuilderX.app/Contents/MacOS/HBuilderX": "/HBuilderX.app/Contents/MacOS/HBuilderX",
  "/Atom.app/Contents/MacOS/Atom": "atom",
  "/Atom Beta.app/Contents/MacOS/Atom Beta": "/Atom Beta.app/Contents/MacOS/Atom Beta",
  "/Brackets.app/Contents/MacOS/Brackets": "brackets",
  "/Sublime Text.app/Contents/MacOS/Sublime Text": "/Sublime Text.app/Contents/SharedSupport/bin/subl",
  "/Sublime Text.app/Contents/MacOS/sublime_text": "/Sublime Text.app/Contents/SharedSupport/bin/subl",
  "/Sublime Text 2.app/Contents/MacOS/Sublime Text 2": "/Sublime Text 2.app/Contents/SharedSupport/bin/subl",
  "/Sublime Text Dev.app/Contents/MacOS/Sublime Text": "/Sublime Text Dev.app/Contents/SharedSupport/bin/subl",
  "/PhpStorm.app/Contents/MacOS/phpstorm": "/PhpStorm.app/Contents/MacOS/phpstorm",
  "/PyCharm.app/Contents/MacOS/pycharm": "/PyCharm.app/Contents/MacOS/pycharm",
  "/PyCharm CE.app/Contents/MacOS/pycharm": "/PyCharm CE.app/Contents/MacOS/pycharm",
  "/IntelliJ IDEA.app/Contents/MacOS/idea": "/IntelliJ IDEA.app/Contents/MacOS/idea",
  "/IntelliJ IDEA Ultimate.app/Contents/MacOS/idea": "/IntelliJ IDEA Ultimate.app/Contents/MacOS/idea",
  "/IntelliJ IDEA Community Edition.app/Contents/MacOS/idea": "/IntelliJ IDEA Community Edition.app/Contents/MacOS/idea",
  "/Zed.app/Contents/MacOS/zed": "zed",
  "/GoLand.app/Contents/MacOS/goland": "/GoLand.app/Contents/MacOS/goland",
  "/AppCode.app/Contents/MacOS/appcode": "/AppCode.app/Contents/MacOS/appcode",
  "/CLion.app/Contents/MacOS/clion": "/CLion.app/Contents/MacOS/clion",
  "/RubyMine.app/Contents/MacOS/rubymine": "/RubyMine.app/Contents/MacOS/rubymine",
  "/MacVim.app/Contents/MacOS/MacVim": "mvim",
  "/Rider.app/Contents/MacOS/rider": "/Rider.app/Contents/MacOS/rider"
}, U = {
  kiro: ["/Kiro.app/Contents/MacOS/Electron"],
  cursor: ["/Cursor.app/Contents/MacOS/Cursor"],
  comate: ["/Comate.app/Contents/MacOS/Electron"],
  qoder: [
    "/Qoder.app/Contents/MacOS/Electron",
    "/Qoder CN.app/Contents/MacOS/Electron"
  ],
  windsurf: ["/Windsurf.app/Contents/MacOS/Electron"],
  trae: [
    "/Trae.app/Contents/MacOS/Electron",
    "/Trae CN.app/Contents/MacOS/Electron"
  ],
  codebuddy: [
    "/Applications/CodeBuddy.app/Contents/MacOS/Electron",
    "/Applications/CodeBuddy CN.app/Contents/MacOS/Electron"
  ],
  antigravity: ["Antigravity.app/Contents/MacOS/Electron"],
  code: ["/Visual Studio Code.app/Contents/MacOS/Electron", "/Visual Studio Code.app/Contents/MacOS/Code"],
  "code-insiders": [
    "/Visual Studio Code - Insiders.app/Contents/MacOS/Electron"
  ],
  webstorm: ["/WebStorm.app/Contents/MacOS/webstorm"],
  atom: ["/Atom.app/Contents/MacOS/Atom"],
  hbuilder: ["/HBuilderX.app/Contents/MacOS/HBuilderX"],
  phpstorm: ["/PhpStorm.app/Contents/MacOS/phpstorm"],
  pycharm: ["/PyCharm.app/Contents/MacOS/pycharm"],
  idea: ["/IntelliJ IDEA.app/Contents/MacOS/idea"],
  codium: ["/VSCodium.app/Contents/MacOS/Electron"],
  goland: ["/GoLand.app/Contents/MacOS/goland"],
  colin: ["/CLion.app/Contents/MacOS/clion"],
  appcode: ["/AppCode.app/Contents/MacOS/appcode"],
  "atom-beta": ["/Atom Beta.app/Contents/MacOS/Atom Beta"],
  brackets: ["/Brackets.app/Contents/MacOS/Brackets"],
  rider: ["/Rider.app/Contents/MacOS/rider"],
  rubymine: ["/RubyMine.app/Contents/MacOS/rubymine"],
  sublime: ["/Sublime Text.app/Contents/MacOS/sublime_text"],
  zed: ["/Zed.app/Contents/MacOS/zed"]
}, A = {
  kiro: "kiro",
  cursor: "cursor",
  comate: "comate",
  qoder: "qoder",
  windsurf: "windsurf",
  trae: "trae",
  codebuddy: "codebuddy",
  antigravity: "antigravity",
  code: "vscode",
  codium: "vscodium"
}, q = {
  kiro: "kiro",
  antigravity: "antigravity",
  cursor: "cursor",
  windsurf: "windsurf",
  code: "code",
  vscodium: "vscodium",
  codium: "codium",
  webstorm: "webstorm",
  "webstorm.sh": "webstorm",
  hbuilderx: "hbuilderx",
  "hbuilderx.sh": "hbuilderx",
  atom: "atom",
  Brackets: "brackets",
  "code-insiders": "code-insiders",
  emacs: "emacs",
  gvim: "gvim",
  idea: "idea",
  "idea.sh": "idea",
  phpstorm: "phpstorm",
  "phpstorm.sh": "phpstorm",
  pycharm: "pycharm",
  "pycharm.sh": "pycharm",
  rubymine: "rubymine",
  "rubymine.sh": "rubymine",
  sublime_text: "subl",
  vim: "vim",
  goland: "goland",
  "goland.sh": "goland",
  rider: "rider",
  "rider.sh": "rider"
}, z = {
  kiro: ["kiro"],
  antigravity: ["antigravity"],
  code: ["code"],
  "code-insiders": ["code-insiders"],
  webstorm: ["webstorm", "webstorm.sh"],
  cursor: ["cursor"],
  windsurf: ["windsurf"],
  trae: ["trae"],
  comate: ["comate"],
  qoder: ["qoder"],
  codebuddy: ["codebuddy"],
  atom: ["atom"],
  hbuilder: ["hbuilderx", "hbuilderx.sh"],
  phpstorm: ["phpstorm", "phpstorm.sh"],
  pycharm: ["pycharm", "pycharm.sh"],
  idea: ["idea", "idea.sh"],
  codium: ["vscodium"],
  goland: ["goland"],
  brackets: ["Brackets"],
  rider: ["rider"],
  rubymine: ["rubymine", "rubymine.sh"],
  sublime: ["sublime_text"],
  vim: ["vim"],
  emacs: ["emacs"]
}, K = {
  "Kiro.exe": "",
  "Antigravity.exe": "",
  "Cursor.exe": "",
  "Windsurf.exe": "",
  "Trae.exe": "",
  "Trae CN.exe": "",
  "comate.exe": "",
  "Qoder.exe": "",
  "Qoder CN.exe": "",
  "CodeBuddy.exe": "",
  "CodeBuddy CN.exe": "",
  "Code.exe": "",
  "Code - Insiders.exe": "",
  "VSCodium.exe": "",
  "webstorm.exe": "",
  "webstorm64.exe": "",
  "HBuilderX.exe": "",
  "HBuilderX64.exe": "",
  "HBuilder.exe": "",
  "HBuilder64.exe": "",
  "Brackets.exe": "",
  "atom.exe": "",
  "sublime_text.exe": "",
  "notepad++.exe": "",
  "clion.exe": "",
  "clion64.exe": "",
  "idea.exe": "",
  "idea64.exe": "",
  "phpstorm.exe": "",
  "phpstorm64.exe": "",
  "pycharm.exe": "",
  "pycharm64.exe": "",
  "rubymine.exe": "",
  "rubymine64.exe": "",
  "goland.exe": "",
  "goland64.exe": "",
  "rider.exe": "",
  "rider64.exe": ""
}, G = {
  kiro: ["Kiro.exe"],
  antigravity: ["Antigravity.exe"],
  code: ["Code.exe"],
  "code-insiders": ["Code - Insiders.exe"],
  webstorm: ["webstorm.exe", "webstorm64.exe"],
  cursor: ["Cursor.exe"],
  windsurf: ["Windsurf.exe"],
  trae: ["Trae.exe", "Trae CN.exe"],
  comate: ["comate.exe"],
  qoder: ["Qoder.exe", "Qoder CN.exe"],
  codebuddy: ["CodeBuddy.exe", "CodeBuddy CN.exe"],
  atom: ["atom.exe"],
  hbuilder: [
    "HBuilderX.exe",
    "HBuilder.exe",
    "HBuilderX64.exe",
    "HBuilder64.exe"
  ],
  phpstorm: ["phpstorm.exe", "phpstorm64.exe"],
  pycharm: ["pycharm.exe", "pycharm64.exe"],
  idea: ["idea.exe", "idea64.exe"],
  codium: ["VSCodium.exe"],
  goland: ["goland.exe", "goland64.exe"],
  colin: ["clion.exe", "clion64.exe"],
  brackets: ["Brackets.exe"],
  rider: ["rider.exe", "rider64.exe"],
  rubymine: ["rubymine.exe", "rubymine64.exe"],
  sublime: ["sublime_text.exe"],
  notepad: ["notepad++.exe"]
}, R = {
  darwin: J,
  linux: q,
  win32: K
}, _ = {
  darwin: U,
  linux: z,
  win32: G
}, m = "{file}", u = "{line}", g = "{column}";
function Z(t, e, n, o) {
  let a = "".concat(t, ":").concat(e, ":").concat(n);
  if (typeof o == "string")
    a = o.replace(m, t).replace(u, e.toString()).replace(g, n.toString());
  else if (o instanceof Array)
    return o.map((r) => r.replace(m, t).replace(u, e.toString()).replace(g, n.toString()));
  return [a];
}
const Y = ["-g", "-r", "{file}:{line}:{column}"];
function ee(t) {
  const {
    editorBasename: e,
    fileName: n,
    lineNumber: o,
    colNumber: a,
    workspace: r,
    openWindowParams: i,
    pathFormat: c
  } = t, x = ne({ editorBasename: e, openWindowParams: i, workspace: r }) || Y;
  return Z(n, o, a, c || x);
}
function te(t) {
  let e = v.basename(t).replace(/\.(exe|cmd|bat|sh)$/i, "");
  const n = process.platform, o = Object.keys(_[n]);
  for (let a = 0; a < o.length; a++)
    if ((_[n][o[a]] || []).some((i) => t.endsWith(i))) {
      e = o[a];
      break;
    }
  return e.toLowerCase();
}
function ne(t) {
  const { editorBasename: e, openWindowParams: n, workspace: o } = t;
  switch (e) {
    case "atom":
    case "atom beta":
    case "subl":
    case "sublime":
    case "sublime_text":
    case "wstorm":
    case "charm":
    case "zed":
      return "".concat(m, ":").concat(u, ":").concat(g);
    case "notepad++":
      return ["-n" + u, "-c" + g, m];
    case "vim":
    case "mvim":
      return ["+call cursor(".concat(u, ", ").concat(g, ")"), m];
    case "joe":
    case "gvim":
      return ["+" + u, m];
    case "emacs":
    case "emacsclient":
      return ["+" + u + ":" + g, m];
    case "rmate":
    case "mate":
    case "mine":
      return ["--line", u, m];
    case "code":
    case "code-insiders":
    case "code - insiders":
    case "codium":
    case "cursor":
    case "windsurf":
    case "trae":
    case "codebuddy":
    case "antigravity":
    case "comate":
    case "qoder":
    case "vscodium":
    case "hbuilderx":
    case "hbuilder":
    case "kiro":
      return [
        ...o ? [o] : [],
        "-g",
        ...n ? [n] : [],
        "".concat(m, ":").concat(u, ":").concat(g)
      ];
    case "appcode":
    case "clion":
    case "clion64":
    case "idea":
    case "idea64":
    case "phpstorm":
    case "phpstorm64":
    case "pycharm":
    case "pycharm64":
    case "rubymine":
    case "rubymine64":
    case "webstorm":
    case "webstorm64":
    case "goland":
    case "goland64":
    case "rider":
    case "rider64":
      return [
        ...o ? [o] : [],
        "--line",
        u,
        m
      ];
  }
  return "";
}
function oe(t) {
  let e = {};
  process.env && (e = E({}, process.env));
  let n = "";
  if (t) {
    const o = v.resolve(t, ".env.local");
    T.existsSync(o) && (n = o);
  }
  if (!n) {
    const o = v.resolve(process.cwd(), ".env.local");
    T.existsSync(o) && (n = o);
  }
  if (n) {
    const o = T.readFileSync(n, "utf-8"), a = j.parse(o || "");
    e = E(E({}, a), e);
  }
  return e;
}
function $(t, e) {
  return oe(e)[t] || null;
}
const re = {
  darwin: "ps ax -o comm=",
  linux: "ps -eo comm --sort=comm",
  win32: 'powershell -NoProfile -Command "Get-CimInstance -Query \\"select executablepath from win32_process where executablepath is not null\\" | % { $_.ExecutablePath }"'
};
function ae(t, e, n) {
  let o = null;
  const a = $("CODE_EDITOR", e || "");
  if (a) {
    const r = D(a);
    if (r)
      o = r;
    else
      return [a];
  }
  if (!o && t) {
    const r = D(t);
    r && (o = r);
  }
  if (!o && n) {
    const r = se();
    if (r)
      return [r];
  }
  try {
    let r;
    const i = process.platform, c = process.platform === "win32", h = re[i], x = R[i];
    L(c);
    const I = w.execSync(h, { encoding: "utf-8" }), C = Object.keys(x), S = I.split(c ? "\r\n" : "\n").map((b) => b.trim());
    for (let b = 0; b < C.length; b++) {
      const p = C[b];
      let s = "", M = "";
      if (c) {
        const f = S.find(
          (y) => v.basename(y).toLowerCase() === p.toLowerCase()
        );
        f && (M = v.basename(f), s = f);
      } else if (i === "darwin") {
        const f = S.find(
          (y) => y.toLowerCase().endsWith(p.toLowerCase())
        );
        if (f) {
          const y = f.replace(p, ""), d = x[p];
          M = p, d.includes("/") ? s = "".concat(y).concat(d) : s = d;
        }
      } else
        I.indexOf(p) !== -1 && (M = p, s = x[p]);
      if (M && s) {
        if (o != null && o.includes(M))
          return [s];
        r || (r = [s]);
      }
    }
    if (r)
      return r;
  } catch (r) {
  }
  return process.env.VISUAL ? [process.env.VISUAL] : process.env.EDITOR ? [process.env.EDITOR] : [null];
}
const D = (t) => {
  const e = process.platform;
  return _[e] && _[e][t] || null;
}, L = (t) => {
  if (t)
    try {
      w.execSync("chcp 65001");
    } catch (e) {
    }
};
function se() {
  const t = process.platform, e = Object.keys(R[t]);
  try {
    return ce(process.pid, t, e);
  } catch (n) {
    return console.log(l.red("Error while getting editor by PID:"), n), null;
  }
}
function ce(t, e, n) {
  let o = 0;
  const a = 50, r = 0;
  for (; t && t !== r && o < a; ) {
    const i = ie(t, e);
    if (!i)
      break;
    const { command: c, parentPid: h } = i;
    if (me(c, n))
      return c;
    t = h, o++;
  }
  return null;
}
function ie(t, e) {
  switch (e) {
    case "darwin":
    case "linux":
      return pe(t);
    case "win32":
      return de(t);
    default:
      return null;
  }
}
function pe(t) {
  try {
    const n = w.execSync("ps -p ".concat(t, " -o ppid=,comm="), {
      encoding: "utf8"
    }).trim().split("\n");
    return n.length ? le(n[0].trim()) : null;
  } catch (e) {
    return null;
  }
}
function de(t) {
  try {
    L(!0);
    const e = 'powershell -NoProfile -Command "Get-CimInstance -ClassName Win32_Process -Filter "ProcessId='.concat(t, '" | Select-Object ParentProcessId, ExecutablePath | ConvertTo-Csv -NoTypeInformation'), o = w.execSync(e, {
      encoding: "utf8"
    }).trim().split("\r\n").filter((i) => i.trim()).slice(1);
    if (o.length === 0)
      return null;
    const [a, r] = o[0].split(",");
    return !r || !a ? null : {
      command: r.slice(1, -1),
      parentPid: Number(a.slice(1, -1))
    };
  } catch (e) {
    return null;
  }
}
function le(t) {
  const e = t.match(/^(\d+)\s+(.+)$/);
  return e ? {
    command: e[2],
    parentPid: parseInt(e[1])
  } : null;
}
function me(t, e) {
  return e.some(
    (n) => t.toLowerCase().endsWith(n.toLowerCase())
  );
}
function ue(t) {
  switch (t) {
    case "vim":
    case "emacs":
    case "nano":
      return !0;
  }
  return !1;
}
function Ce(t) {
  const e = $(
    "CODE_INSPECTOR_FORMAT_PATH",
    t
  );
  if (e)
    try {
      return JSON.parse(e);
    } catch (n) {
      return null;
    }
  return null;
}
function k(t, e) {
  console.log(
    l.red("Could not open " + v.basename(t) + " in the editor.")
  ), e && (e[e.length - 1] !== "." && (e += "."), console.log(
    l.red("The editor process exited with an error: " + e)
  )), console.log(
    "To set up the editor integration, add something like " + l.cyan("CODE_EDITOR=code") + " to the " + l.green(".env.local") + " file in your project folder, or add " + l.green('editor: "code"') + " to CodeInspectorPlugin config, and then restart the development server. Learn more: " + l.green("https://goo.gl/MMTaZt")
  );
}
let O = null;
function Se(t) {
  return t === "reuse" ? "-r" : t === "new" ? "-n" : "";
}
function Ee(t) {
  let {
    file: e,
    line: n = 1,
    column: o = 1,
    editor: a,
    method: r,
    format: i,
    onError: c,
    rootDir: h,
    usePid: x,
    type: I = "exec"
  } = t;
  if (!T.existsSync(e))
    return;
  let [C, ...S] = ae(a, h, x);
  const b = Ce(h || "") || i;
  if (!C || C.toLowerCase() === "none") {
    typeof c == "function" ? c(e, "Failed to recognize IDE automatically") : console.log(
      "Failed to recognize IDE automatically, add something like " + l.cyan("CODE_EDITOR=code") + " to the " + l.green(".env.local") + " file in your project folder, or add " + l.green('editor: "code"') + " to CodeInspectorPlugin config, and then restart the development server. Learn more: " + l.green("https://goo.gl/MMTaZt")
    );
    return;
  }
  const p = te(
    C
  );
  if (I === "open" && process.platform === "darwin" && A[p])
    O = w.spawn(
      "open",
      ["".concat(A[p], "://file").concat(e, ":").concat(n, ":").concat(o)],
      {
        stdio: "ignore",
        env: P(E({}, process.env), {
          NODE_OPTIONS: ""
        })
      }
    );
  else {
    process.platform === "linux" && e.startsWith("/mnt/") && /Microsoft/i.test(Q.release()) && (e = v.relative("", e));
    let s = null;
    if (n ? S = S.concat(
      ee({
        editorBasename: p,
        fileName: e,
        lineNumber: n,
        colNumber: o,
        workspace: s,
        openWindowParams: Se(r),
        pathFormat: b
      })
    ) : S.push(e), O && ue(C) && O.kill("SIGKILL"), process.platform === "win32") {
      const M = (d) => d.replace(/([&|<>,;=^])/g, "^$1"), f = (d) => d.includes("^") ? '^"'.concat(d, '^"') : d.includes(" ") ? '"'.concat(d, '"') : d, y = [C, ...S.map(M)].map(f).join(" ");
      O = w.exec(y, {
        stdio: "ignore",
        // @ts-ignore
        shell: !0,
        env: P(E({}, process.env), {
          NODE_OPTIONS: ""
        })
      });
    } else
      O = w.spawn(C, S, {
        stdio: "ignore",
        env: P(E({}, process.env), {
          NODE_OPTIONS: ""
        })
      });
  }
  O.on("exit", function(s) {
    O = null, s && (typeof c == "function" ? c(e, "(code " + s + ")") : k(e, "(code " + s + ")"));
  }), O.on("error", function(s) {
    typeof c == "function" ? c(e, s.message) : k(e, s.message);
  });
}
export {
  Z as formatOpenPath,
  $ as getEnvVariable,
  oe as getEnvVariables,
  Ee as launchIDE
};
